File: ios/Modules/SwiftUIComponents/Sources/SwiftUIComponents/Components/NextEpisode/NextEpisodesSection.swift
import SwiftUI

public struct NextEpisodesSection: View {
    @Theme private var theme

    private let title: String
    private let episodes: [SwiftNextEpisode]
    private let chevronStyle: ChevronStyle
    private let onEpisodeClick: (SwiftNextEpisode) -> Void
    private let onSeeAllClick: () -> Void

    public init(
        title: String,
        episodes: [SwiftNextEpisode],
        chevronStyle: ChevronStyle = .none,
        onEpisodeClick: @escaping (SwiftNextEpisode) -> Void,
        onSeeAllClick: @escaping () -> Void = {}
    ) {
        self.title = title
        self.episodes = episodes
        self.chevronStyle = chevronStyle
        self.onEpisodeClick = onEpisodeClick
        self.onSeeAllClick = onSeeAllClick
    }

    public var body: some View {
        if !episodes.isEmpty {
            VStack {
                ChevronTitle(
                    title: title,
                    chevronStyle: chevronStyle,
                    action: onSeeAllClick
                )

                ScrollView(.horizontal, showsIndicators: false) {
                    HStack {
                        ForEach(episodes, id: \.episodeId) { episode in
                            NextEpisodeCard(
                                episode: episode,
                                onEpisodeClick: { onEpisodeClick(episode) }
                            )
                            .padding([.leading, .trailing], theme.spacing.xxSmall + 2)
                            .padding(.leading, episode.episodeId == episodes.first?.episodeId ? theme.spacing.small - 2 : 0)
                            .padding(.trailing, episode.episodeId == episodes.last?.episodeId ? theme.spacing.xSmall : 0)
                        }
                    }
                }
            }
            .padding(.bottom, theme.spacing.medium)
        }
    }
}

#Preview {
    VStack {
        NextEpisodesSection(
            title: "Up Next",
            episodes: [
                SwiftNextEpisode(
                    showTraktId: 123,
                    showName: "The Walking Dead: Daryl Dixon",
                    imageUrl: "https://image.tmdb.org/t/p/w780/ydlY3iPfeOAvu8gVqrxPoMvzNCn.jpg",
                    episodeId: 456,
                    episodeTitle: "L'âme Perdue",
                    episodeNumber: "S02E01",
                    runtime: "45 min",
                    overview: "Daryl washes ashore in France and struggles to piece together how he got there and why.",
                    badge: .premiere
                ),
                SwiftNextEpisode(
                    showTraktId: 124,
                    showName: "Wednesday",
                    imageUrl: "https://image.tmdb.org/t/p/w780/dC0oTEMAPnBzM0RU15qArHoqnH5.jpg",
                    episodeId: 789,
                    episodeTitle: "Wednesday's Child Is Full of Woe",
                    episodeNumber: "S02E02",
                    runtime: "50 min",
                    overview: "Wednesday arrives at Nevermore Academy.",
                    badge: .new
                ),
            ],
            chevronStyle: .chevronOnly,
            onEpisodeClick: { _ in }
        )
    }
}

File: features/discover/ui/src/main/java/com/thomaskioko/tvmaniac/discover/ui/component/NextEpisodesSection.kt
package com.thomaskioko.tvmaniac.discover.ui.component

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.tooling.preview.PreviewWrapper
import androidx.compose.ui.unit.dp
import com.thomaskioko.tvmaniac.compose.components.BoxTextItems
import com.thomaskioko.tvmaniac.compose.components.ThemePreviews
import com.thomaskioko.tvmaniac.compose.components.TvManiacPreviewWrapperProvider
import com.thomaskioko.tvmaniac.discover.presenter.model.NextEpisodeUiModel
import com.thomaskioko.tvmaniac.i18n.MR.strings.str_more
import com.thomaskioko.tvmaniac.i18n.resolve
import com.thomaskioko.tvmaniac.testtags.discover.DiscoverTestTags
import dev.chrisbanes.snapper.rememberSnapperFlingBehavior
import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.persistentListOf

@Composable
internal fun NextEpisodesSection(
    title: String,
    modifier: Modifier = Modifier,
    nextEpisodes: ImmutableList<NextEpisodeUiModel>,
    onEpisodeClick: (NextEpisodeUiModel) -> Unit,
    onSeeAllClick: () -> Unit = {},
) {
    AnimatedVisibility(
        visible = nextEpisodes.isNotEmpty(),
        modifier = modifier,
    ) {
        Column {
            BoxTextItems(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(start = 16.dp),
                title = title,
                label = str_more.resolve(LocalContext.current),
                onMoreClicked = onSeeAllClick,
            )

            val lazyListState = rememberLazyListState()

            LazyRow(
                state = lazyListState,
                flingBehavior = rememberSnapperFlingBehavior(lazyListState),
                contentPadding = PaddingValues(horizontal = 16.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                itemsIndexed(
                    items = nextEpisodes,
                    key = { _, episode -> "next_episode_${episode.showTraktId}_${episode.episodeId}" },
                ) { _, episode ->
                    NextEpisodeCard(
                        modifier = Modifier.testTag(DiscoverTestTags.upNextCard(episode.showTraktId)),
                        episode = episode,
                        onClick = { onEpisodeClick(episode) },
                    )
                }
            }
        }
    }
}

@ThemePreviews
@PreviewWrapper(TvManiacPreviewWrapperProvider::class)
@Composable
private fun NextEpisodesSectionPreview() {
    NextEpisodesSection(
        title = "Up Next",
        nextEpisodes = persistentListOf(
            NextEpisodeUiModel(
                showTraktId = 1L,
                showName = "The Walking Dead: Daryl Dixon",
                imageUrl = "/still1.jpg",
                episodeId = 123L,
                episodeTitle = "L'âme Perdue",
                episodeNumberFormatted = "S02E01",
                seasonId = 1L,
                seasonNumber = 2,
                episodeNumber = 1,
                runtime = "45 min",
                overview = "Daryl washes ashore in France and struggles to piece together how he got there and why.",
                isNew = true,
            ),
            NextEpisodeUiModel(
                showTraktId = 2L,
                showName = "Wednesday",
                imageUrl = "/still1.jpg",
                episodeId = 124L,
                episodeTitle = "Wednesday's Child Is Full of Woe",
                episodeNumberFormatted = "S02E02",
                seasonId = 2L,
                seasonNumber = 2,
                episodeNumber = 2,
                runtime = "50 min",
                overview = "Wednesday arrives at Nevermore Academy and immediately gets off on the wrong foot.",
                isNew = false,
            ),
            NextEpisodeUiModel(
                showTraktId = 3L,
                showName = "House of the Dragon",
                imageUrl = "/still1.jpg",
                episodeId = 125L,
                episodeTitle = "The Heirs of the Dragon",
                episodeNumberFormatted = "S03E01",
                seasonId = 3L,
                seasonNumber = 3,
                episodeNumber = 1,
                runtime = "66 min",
                overview = "King Viserys hosts a tournament to celebrate the birth of his second child.",
                isNew = true,
            ),
        ),
        onEpisodeClick = {},
    )
}
